plugins/single_plugins/preserve-output.cpp

```diff
@@ -0,0 +1,314 @@
+            }
+
+            // Sorts with the views closest to front last
+            std::sort(views.begin(), views.end(), last_output_info_z_sort());
+
+            // Move views to this output
+            for (auto& view : views)
+            {
+                auto last_output_info = view_get_data(view);
+                LOGD("Restoring view: ",
+                    view->get_title(), " to: ", output->to_string());
+
+                wf::get_core().move_view_to_output(view, output, false);
+                view->set_fullscreen(last_output_info->fullscreen);
+                view->set_minimized(last_output_info->minimized);
+                view->set_sticky(last_output_info->sticky);
+                view->set_tiled(last_output_info->tiled_edges);
+                if (last_output_info->tiled_edges != 0)
+                {
+                    view->tile_request(last_output_info->tiled_edges);
+                }
+
+                // This is to ensure the next resize is actually done, helps ensure
+                // apps redraw themselves too
+                view->resize(last_output_info->geometry.width + 1,
+                    last_output_info->geometry.height);
+                view->set_geometry(last_output_info->geometry);
+
+                // Focus
+                if (last_output_info->focused)
+                {
+                    LOGD("Focusing view: ", view->get_title());
+                    output->focus_view(view, false);
+                }
+
+                // Z Order
+                output->workspace->bring_to_front(view);
+
+                // Remove all last output info from views
+                view_erase_data(view);
+            }
+        }
+
+        // Start listening for view resize events AFTER this callback has finished
+        output->connect_signal("view_move_request_signal", &view_moved);
+    };
+
+    wf::signal_connection_t view_moved = [=] (wf::signal_data_t *data)
+    {
+        // Triggered whenever a view on this output's geometry changed
+        auto signal_data = (wf::view_move_request_signal*)data;
+        auto view = signal_data->view;
+
+        if (view_has_data(view))
+        {
+            // Currently triggered when view moved from DP-1 to DP-2
+            if (outputs_being_removed.size() == 0)
+            {
+                LOGD("Geometry actually changed, deleting last output info");
+                view_erase_data(view);
+            }
+        }
+    };
+
+  public:
+    void init() override
+    {
+        if (wlr_output_is_noop(output->handle))
+        {
+            // Don't do anything for NO-OP outputs
+            return;
+        }
+
+        wf::get_core().output_layout->connect_signal("output-added",
+            &restore_views_to_output);
+        wf::get_core().output_layout->connect_signal("output-pre-remove",
+            &output_pre_remove);
+        wf::get_core().output_layout->connect_signal("output-removed",
+            &output_removed);
+    }
```

Comment:
If I do that, won't all the data on views be lost when they're in the NOOP-1 output? The plugin is destroyed when the outputs are destroyed, and the data in views and core needs to outlive them.

Reply:
the noop output will also have an instance of the plugins, so it will keep the data alive (because there would be at least once instance of the plugin active). core guarantees that this NOOP output will be created (with the plugins on it) before destroying the other outputs.

Reply:
Got it. Will add this. I'm still a bit unclear - what situation involves all outputs, including NOOP, being destroyed, other than Wayfire shutting down (which would release all memory it's using)?

Reply:
Oh, this is if the user disables the plugin?